services/brig/src/Brig/API/MLS/KeyPackages/Validation.hs

```diff
@@ -0,0 +1,136 @@
+  (kp, tbs) <- parseKeyPackage kpd
+  -- get ciphersuite
+  cs <-
+    maybe
+      (throwErrorDescription (mlsProtocolError "Unsupported ciphersuite"))
+      pure
+      $ cipherSuiteTag (kpCipherSuite (kpTBS kp))
+  -- validate signature
+  -- FUTUREWORK: authenticate signature key
+  let key = bcSignatureKey (kpCredential (kpTBS kp))
+  unless (csVerifySignature cs key (LBS.toStrict tbs) (kpSignature kp)) $
+    throwErrorDescription (mlsProtocolError "Invalid signature")
+  -- validate credential and extensions
+  validateKeyPackage identity kp
+  pure (kpRef cs kpd, kpd)
+
+-- | Parse a key package, and return parsed structure and signed data.
+parseKeyPackage :: KeyPackageData -> Handler r (KeyPackage, LByteString)
+parseKeyPackage (kpData -> kpd) = do
+  (kp, off) <- either (throwErrorDescription . mlsProtocolError) pure (decodeMLSWith kpSigOffset kpd)
+  pure (kp, LBS.take off kpd)
+
+validateKeyPackage :: ClientIdentity -> KeyPackage -> Handler r ()
+validateKeyPackage identity (kpTBS -> kp) = do
+  maybe
+    (throwErrorDescription (mlsProtocolError "Unsupported protocol version"))
+    pure
+    (pvTag (kpProtocolVersion kp) >>= guard . (== ProtocolMLS10))
+  validateCredential identity (kpCredential kp)
+  validateExtensions (kpExtensions kp)
+
+validateCredential :: ClientIdentity -> Credential -> Handler r ()
+validateCredential identity cred = do
+  identity' <-
+    either (throwErrorDescription . mlsProtocolError) pure $
+      decodeMLS' (bcIdentity cred)
+  when (identity /= identity') $
+    throwErrorDescriptionType @MLSIdentityMismatch
+
+data RequiredExtensions f = RequiredExtensions
+  { reLifetime :: f Lifetime,
+    reCapabilities :: f ()
+  }
+
+deriving instance (Show (f Lifetime), Show (f ())) => Show (RequiredExtensions f)
+
+instance Alternative f => Semigroup (RequiredExtensions f) where
+  RequiredExtensions lt1 cap1 <> RequiredExtensions lt2 cap2 =
+    RequiredExtensions (lt1 <|> lt2) (cap1 <|> cap2)
+
+instance Alternative f => Monoid (RequiredExtensions f) where
+  mempty = RequiredExtensions empty empty
+
+checkRequiredExtensions :: Applicative f => RequiredExtensions f -> f (RequiredExtensions Identity)
+checkRequiredExtensions re =
+  RequiredExtensions
+    <$> (Identity <$> reLifetime re)
+    <*> (Identity <$> reCapabilities re)
+
+findExtensions :: [SomeExtension] -> Maybe (RequiredExtensions Identity)
+findExtensions = checkRequiredExtensions . foldMap findExtension
+
+findExtension :: SomeExtension -> RequiredExtensions Maybe
+findExtension (SomeExtension SLifetimeExtensionTag lt) = RequiredExtensions (pure lt) Nothing
+findExtension (SomeExtension SCapabilitiesExtensionTag _) = RequiredExtensions Nothing (pure ())
+findExtension _ = mempty
+
+validateExtensions :: [Extension] -> Handler r ()
+validateExtensions exts = do
+  re <-
+    maybe (throwErrorDescription (mlsProtocolError "Missing required extensions")) pure $
+      findExtensions (foldMap (maybeToList . decodeExtension) exts)
```

Comment:
Does this silently swallow invalid extensions in a list with at least one valid extension, i.e., is that fine?

Also, to me the error message on line 117 does not seem to match what the code is checking, but I could be wrong.

Reply:
Invalid extensions (i.e. extensions with the reserved extension tag) were indeed swallowed. Not sure how important this is, to be honest, but it's fixed now.

I'm not sure what you mean by the error not matching, can you elaborate?

Reply:
It must be me not understanding this part of the code well. Please ignore my rambling!